src/csrf.test.js

```diff
@@ -0,0 +1,228 @@
+import assert from 'assert';
+import sinon from 'sinon';
+
+import _ from 'underscore';
+import Backbone from 'backbone';
+import Cookies from 'js-cookie';
+
+import wrapWithCSRF from './csrf.js';
+
+var myHeader = 'XYZ', myCookie = 'seasurfnugget', myToken = 'xyz123';
+
+describe('wrapWithCSRF', function() {
+    it('refuses zero argument calls', function() {
+        assert.throws(wrapWithCSRF, TypeError);
+    });
+
+    it('refuses single argument calls', function() {
+        assert.throws(_.partial(wrapWithCSRF, _.noop), TypeError);
+    });
+
+    it('refuses two argument calls', function() {
+        assert.throws(_.partial(wrapWithCSRF, _.noop, myHeader), TypeError);
+    });
+
+    it('accepts three argument calls', function() {
+        var wrapped = wrapWithCSRF(_.noop, myHeader, myCookie);
+        assert(_.isFunction(wrapped));
+    });
+
+    it('is unaffected by a fourth argument', function() {
+        var wrapped = wrapWithCSRF(_.noop, myHeader, myCookie, false);
+        assert(_.isFunction(wrapped));
+    });
+
+    it('refuses non-functions as first argument', function() {
+        var faulty = _.partial(wrapWithCSRF, myHeader, myHeader, myCookie)
+        assert.throws(faulty, TypeError);
+    });
+
+    it('defaults to Backbone.sync if the first argument is null', function() {
+        sinon.spy(Backbone, 'sync');
+        var wrapped = wrapWithCSRF(null, myHeader, myCookie);
+        assert(_.isFunction(wrapped));
+        // We expect an exception because we don't tell Backbone.sync which URL
+        // to visit.
+        assert.throws(wrapped);
+        assert(Backbone.sync.calledOnce);
+        Backbone.sync.restore();
+    });
+
+    var specTextTemplate = _.template(
+        'refuses <%- fault %>strings as <%- ordinal %> argument'
+    );
+    _.each(['second', 'third'], function(ordinal, index) {
```

Comment:
What exactly is the purpose of this?

Reply:
I was going to write four tests, fitting in the previous series of tests, that have almost exactly the same logic, so I wanted to avoid repeating myself. The double loop creates the following four spec descriptions out of the above template:

```
refuses non-strings as second argument
refuses empty strings as second argument
refuses non-strings as third argument
refuses empty strings as third argument
```

I feel that a nested `describe` is not appropriate here, and I also do not think it would help to make the code more explicit. However, I can think of another way to avoid repetition that still enables me to write the spec descriptions in full. I will do that.